extensions/amp-carousel/0.1/test-e2e/test-slidescroll-autoplay.js

```diff
@@ -22,45 +22,53 @@ describes.endtoend(
     environments: ['single'],
   },
   async function(env) {
-    const ActionTrust = {LOW: 1, HIGH: 3};
     let controller;
-    /**
-     * Attach an event listener to page to capture a custom event.
-     * If given a selector, click on it to fire the event being listened for.
-     * @return {!Promise}
-     */
-    function getActionTrustFor(opt_selector) {
-      return controller.driver.executeScript(opt_selector => {
-        return new Promise(resolve => {
-          document.addEventListener('slideChange', function forwardEvent(e) {
-            // Listen once
-            document.removeEventListener('slideChange', forwardEvent);
-            resolve(e.data.actionTrust);
-          });
-          opt_selector ? document.querySelector(opt_selector).click() : null;
-        });
-      }, opt_selector);
-    }
 
     beforeEach(async () => {
       controller = env.controller;
     });
 
     it('should fire low trust event for autoplay advance', async () => {
       for (let i = 0; i < 3; i++) {
-        const actionTrust = await getActionTrustFor(/* autoplay */);
-        await expect(actionTrust).to.equal(ActionTrust.LOW);
+        const actionTrust = await controller.evaluate(() => {
+          return new Promise(resolve => {
+            document.addEventListener('slideChange', function forwardEvent(e) {
+              // Listen once
+              document.removeEventListener('slideChange', forwardEvent);
+              resolve(e.data.actionTrust);
+            });
+          });
+        });
+        await expect(actionTrust).to.equal(1); // ActionTrust.LOW
       }
     });
 
     it('should fire high trust event on user interaction', async () => {
-      const actionTrust = await getActionTrustFor('.amp-carousel-button-next');
-      await expect(actionTrust).to.equal(ActionTrust.HIGH);
+      const actionTrust = await controller.evaluate(() => {
+        return new Promise(resolve => {
+          document.addEventListener('slideChange', function forwardEvent(e) {
+            // Listen once
+            document.removeEventListener('slideChange', forwardEvent);
+            resolve(e.data.actionTrust);
+          });
+          document.querySelector('.amp-carousel-button-next').click();
```

Comment:
Ahh, sorry I just meant to extract this line rather than inline the entire helper. That way the test readers can read the logic of the test naturally like:

1. Click button
2. Wait for slideChange event
3. Assert actionTrust of event is 3

I also wonder if we could set up/tear down the listener in beforeEach/afterEach. If either of those are troublesome, just tweaking the helper a bit would be nice for semantics.

```js
// Before
const actionTrust = await getActionTrustFor('.amp-carousel-button-next');

// After
const event = await slideChangeEventAfterClicking('.amp-carousel-button-next');
expect (event.data.actionTrust).to.equal(3); // ActionTrust.HIGH
```

Reply:
Looks like taking and resolving the same `event` leads to a circular reference, so I had the helper return `event.data` instead. Also used the `once` option when adding the event listener, so should be more readable now.